src/partial-evaluators/BinaryExpression.js

```diff
@@ -9,16 +9,90 @@
 
 /* @flow */
 
-import type { BabelNodeBinaryExpression, BabelNodeStatement } from "babel-types";
+import type { BabelNodeBinaryExpression, BabelNodeExpression, BabelNodeStatement } from "babel-types";
 import type { LexicalEnvironment } from "../environment.js";
 import type { Realm } from "../realm.js";
 
-import { AbruptCompletion } from "../completions.js";
-import { Value } from "../values/index.js";
+import { computeBinary, getPureBinaryOperationResultType } from "../evaluators/BinaryExpression.js";
+import { AbruptCompletion, Completion, PossiblyNormalCompletion } from "../completions.js";
+import { TypesDomain, ValuesDomain } from "../domains/index.js";
+import { composePossiblyNormalCompletions, updatePossiblyNormalCompletionWithValue } from "../methods/index.js";
+import { AbstractValue, BooleanValue, ConcreteValue, NullValue, UndefinedValue, Value } from "../values/index.js";
+
+import * as t from "babel-types";
+import invariant from "../invariant.js";
 
 export default function (
   ast: BabelNodeBinaryExpression, strictCode: boolean, env: LexicalEnvironment, realm: Realm
-): [AbruptCompletion | Value, BabelNodeBinaryExpression, Array<BabelNodeStatement>] {
-  let result = env.evaluateCompletionDeref(ast, strictCode);
-  return [result, ast, []];
+): [Completion | Value, BabelNodeExpression, Array<BabelNodeStatement>] {
+  let [lval, leftAst, leftIO] = env.partiallyEvaluateCompletionDeref(ast.left, strictCode);
+  if (lval instanceof AbruptCompletion) return [lval, (leftAst: any), leftIO];
+  let leftCompletion;
+  if (lval instanceof PossiblyNormalCompletion) {
+    leftCompletion = lval;
+    lval = leftCompletion.value;
+  }
+  invariant(lval instanceof Value);
+
+  let [rval, rightAst, rightIO] = env.partiallyEvaluateCompletionDeref(ast.right, strictCode);
+  let io = leftIO.concat(rightIO);
+  if (rval instanceof AbruptCompletion) {
+    return [rval, t.binaryExpression(ast.operator, (leftAst: any), (rightAst: any)), io];
+  }
+  let rightCompletion;
+  if (rval instanceof PossiblyNormalCompletion) {
+    rightCompletion = rval;
+    rval = rightCompletion.value;
+  }
+  invariant(rval instanceof Value);
+
+  let op = ast.operator;
+  let resultValue, resultAst;
+  if (lval instanceof ConcreteValue) {
+    if (rval instanceof ConcreteValue) {
+      resultValue = computeBinary(realm, op, lval, rval);
+      resultAst = t.valueToNode(resultValue.serialize());
+    }
+  }
+  if (resultValue === undefined && (op === "==" || op === "===" || op === "!=" || op === "!==")) {
+    if (!lval.mightNotBeObject() && (rval instanceof NullValue || rval instanceof UndefinedValue) ||
```

Comment:
I'm following the precedent established by Nikolai in the corresponding Evaluator class, but I also happen to agree with Nikolai that this is bit easier to read and maintain because it would be far too long otherwise.